### DISCUSSION TITLE
How to constantly send a request to the server?

### DISCUSSION THREAD

**Discussion by MarlikAlmighty** [cesanta/mongoose#2810] [Ask questions, get answers] [answered]
Thanks for the library, it's great. I have an amateurish question.

In this example: https://github.com/cesanta/mongoose/blob/master/tutorials/websocket/websocket-client/main.c
 we can only send a message to the server once. How to make it so that it is constantly sent, for example, reading from the console?

Sorry for this question, I'm new to C.

**Comment by scaprile** [comment, marked as answer]
It is not a Mongoose issue if you don't have expertise in C, so please don't open an issue when we have a section for questions.
What you ask is explained in the server tutorial, where it is a common practice

**Reply by MarlikAlmighty** [reply]
> It is not a Mongoose issue if you don't have expertise in C, so please don't open an issue when we have a section for questions. What you ask is explained in the server tutorial, where it is a common practice

Thanks for answering. Unfortunately, I didn’t find a place where you can ask such questions.

For example, in golang I do this easily:

```
package main

import (
	"bufio"
	"crypto/tls"
	"log"
	"net/url"
	"os"

	"github.com/gorilla/websocket"
)

func main() {

	tlsConfig := tls.Config{
		InsecureSkipVerify: true,
	}

	dial := websocket.Dialer{TLSClientConfig: &tlsConfig}

	u := url.URL{Scheme: "wss", Host: "127.0.0.1:8443", Path: "/websocket"}
	log.Printf("connecting to %s", u.String())

	c, _, err := dial.Dial(u.String(), nil)
	if err != nil {
		log.Fatalln(err)
	}
	defer c.Close()

	s := bufio.NewScanner(os.Stdin)
	var message []byte

	for s.Scan() {
		if err = c.WriteMessage(websocket.TextMessage, []byte(s.Text())); err != nil {
			log.Fatalln(err)
		}
		if _, message, err = c.ReadMessage(); err != nil {
			log.Fatalln(err)
		} else {
			log.Println(string(message))
		}
	}
}
```
I noticed that the event occurs once, so it is not clear how to send messages to the socket in the future. If you could briefly clarify how to solve this problem, I would be very grateful to you.

**Reply by scaprile** [reply]
> Thanks for answering. Unfortunately, I didn’t find a place where you can ask such questions.

You are now in such a place.
Linked from our main docs page
![Screenshot from 2024-06-27 11-49-36](https://github.com/cesanta/mongoose/assets/6046251/81d47363-65d5-48d1-b948-351b0ee2ebdb)
Visible at top of the page
![Screenshot from 2024-06-27 11-47-26](https://github.com/cesanta/mongoose/assets/6046251/4191f326-7925-4222-b35d-eeac17fa9be3)

> I noticed that the event occurs once

Of course it occurs once, it is an event ! Mongoose is event-driven.
You can't loop in an event handler, if you need to send something more you have to send timely or catch the MG_EV_POLL event.
Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).
https://mongoose.ws/documentation/tutorials/websocket/websocket-server/#websocket-broadcast
https://mongoose.ws/documentation/#events
> MG_EV_POLL,       // mg_mgr_poll iteration        uint64_t *uptime_millis

https://mongoose.ws/documentation/tutorials/error-handling/#handling-connection-timeouts
https://mongoose.ws/documentation/tutorials/core/multi-threaded/#the-one-to-many-case

**Reply by MarlikAlmighty** [reply]
Thanks for the answer. I found your documentation and read it periodically. I was also able to understand the events. I just printed them into the console)) 

 `You can't loop in an event handler, if you need to send something more you have to send timely or catch the MG_EV_POLL event.`

But this is not clear. I can catch the event, but how do I send from it? After all, reading from stdin blocks. For example, I have this client case:

```
static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data)
{

    // printf("[EV]: %d\n", ev);

    if (ev == MG_EV_CONNECT)
    {
        struct mg_tls_opts opts = {};
        mg_tls_init(c, &opts);
    }
    else if (ev == MG_EV_OPEN)
    {
        printf("OPEN: %d\n", ev);
    }
    else if (ev == MG_EV_WS_OPEN)
    {
        printf("WS OPEN: %d\n", ev);
        // When websocket handshake is successful, send message
        mg_ws_send(c, "hello-server", 25, WEBSOCKET_OP_TEXT);
    }
    else if (ev == MG_EV_POLL)
    {
        printf("POLL: %d\n", ev);
        // IS IT HERE?
    }
    else if (ev == MG_EV_WS_MSG)
    {
        // When we get echo response, print it
        struct mg_ws_message *wm = (struct mg_ws_message *)ev_data;
        printf("ECHO: [%.*s]\n", (int)wm->data.len, wm->data.ptr);
    }
    else if (ev == MG_EV_ERROR)
    {
        // On error, log error message
        MG_ERROR(("[ERROR]: %p %s", c->fd, (char *)ev_data));
    }
    else if (ev == MG_EV_CLOSE)
    {
        *(bool *)fn_data = true; // Signal that we're done
    }
}
```
Nevertheless, thanks for the support and answers.
